pallets/delegation/src/migrations.rs

```diff
@@ -0,0 +1,480 @@
+// be upgradeable to v1, so the value of LATEST_UPGRADEABLE_VERSION would be 0.
+// This needs to be bumped (by 1) every time a new storage migration should take
+// place. Along with this, a new version migrator must be defined and added to
+// the `migrations` vector or the `StorageMigrator`.
+const LATEST_UPGRADEABLE_VERSION: u16 = 0;
+
+// The trait that each version migrator must implement.
+trait VersionMigrator<T: Config> {
+	#[cfg(feature = "try-runtime")]
+	fn pre_migrate(&self) -> Result<(), DelegationMigrationError>;
+	fn migrate(&self) -> Weight;
+	#[cfg(feature = "try-runtime")]
+	fn post_migrate(&self) -> Result<(), DelegationMigrationError>;
+}
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) enum DelegationMigrationError {
+	AlreadyLatest,
+	MigrationResultInconsistent,
+}
+
+// The "manager" of the pallet's storage migrations. It contains a vector of
+// version migrations each of which corresponds to a new runtime upgrade to run,
+// sequencially. It interacts with the `LastUpgradeVersion` of the pallet's
+// storage.
+//
+// When a new runtime upgrade needs to be added, a new component implementing
+// the `VersionMigrator` trait must be defined and added at the end of the
+// `migrations` vector, so that it is always executed after all
+// previous migrations have taken place.
+pub(crate) struct StorageMigrator<T: Config> {
+	// The vector of version migrators.
+	migrations: Vec<Box<dyn VersionMigrator<T>>>,
+}
+
+impl<T: Config> StorageMigrator<T> {
+	pub(crate) fn new() -> Self {
+		Self {
+			migrations: vec![Box::new(v0::V0Migrator {})],
+		}
+	}
+
+	// Checks whether there is at least one storage migration to perform.
+	#[cfg(any(feature = "try-runtime", test))]
+	#[allow(clippy::absurd_extreme_comparisons)]
+	pub(crate) fn pre_migration(&self) -> Result<(), DelegationMigrationError> {
+		ensure!(
+			LastUpgradeVersion::<T>::get() <= migrations::LATEST_UPGRADEABLE_VERSION,
+			DelegationMigrationError::AlreadyLatest
+		);
+
+		Ok(())
+	}
+
+	// It retrieves the latest version deployed on chain, and sequencially applies
+	// any runtime migration until the state is update to the latest version.
+	// It provides a test feature that panics whenver something goes wrong for one
+	// of the version migrators, but in production it is just assumed that all the
+	// migrations will go through.
+	pub(crate) fn migrate(&self) -> Weight {
+		let mut total_weight_used: Weight = 0;
+		let current_version = LastUpgradeVersion::<T>::get();
+		// Applies all the needed migration from the current version up to
+		// LATEST_UPGRADEABLE_VERSION (which is always the latest possible version - 1).
+		for version in current_version..=LATEST_UPGRADEABLE_VERSION {
+			// It is assumed that the are exactly LATEST_UPGRADEABLE_VERSION+1 migrators, so
+			// as long as the pre-condition is met, the following is assumed to never panic.
+			let version_migrator: &dyn VersionMigrator<T> = self.migrations[version as usize].as_ref();
```

Comment:
Instead of assuming, we should rather check with `get(version as usize)`. Moreover, you should ensure this exists in `pre_migrate` to not be able to panic here at all.
```suggestion
			let version_migrator: &dyn VersionMigrator<T> = self.migrations.get(version as usize).unwrap().as_ref();
```

Reply:
I guess there's no difference between [index] and `.get().unwrap()`? We still need to assume it exists via some tests before running the migration. Once we are in the migration, it's already too late, right?

Reply:
That is correct, yes. That's why I would like to check the existence of this in the `pre_migrate` for all versions (see comment [below](https://github.com/KILTprotocol/mashnet-node/pull/226#discussion_r670265414)).